### PR TITLE
[npm-whoami] Add types

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/npm-whoami/npm-whoami-tests.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        // $ExpectType Error | null
+        err;
+
+        // $ExpectType string | undefined
+        username;
+
+        if (err) {
+            username;
+            // ^?
+        }
```
Comment (JoshuaKGoldberg) [inline types/npm-whoami/npm-whoami-tests.ts:14]:
```suggestion
```

File: types/npm-whoami/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    type NpmWhoamiOptions = number | string | NpmWhoamiBothOptions;
+
+    interface NpmWhoamiBothOptions {
+        registry?: string;
+        timeout?: number;
+    }
+
+    function sync(opts: NpmWhoamiOptions): string | undefined;
+}
```
Comment (peterblazejewicz) [inline types/npm-whoami/index.d.ts:24]:
```suggestion
export = whoami;

declare function whoami(callback: whoami.Callback, opts?: whoami.Options): void;

declare namespace whoami {
    interface Callback {
        (err: Error, username: undefined): void;
        (err: null, username: string): void;
    }

    type Options = number | string | BothOptions;

    interface BothOptions {
        registry?: string;
        timeout?: number;
    }

    function sync(opts: Options): string | undefined;
}
```
that way original symbol names could be preserved (notice no change in tests required)

Comment (JoshuaKGoldberg) [reply types/npm-whoami/index.d.ts]: Thanks, this is great!

Just added `Whoami` before `Callback` and `Options` - I've always found it annoying to have to rename constantly...

**Review by sheetalkamat — APPROVED** [review]

**Comment by JoshuaKGoldberg** [comment]
Ready to merge

**Pull request merged**